csharp/src/Ice/Communicator.cs

```diff
@@ -1364,5 +1273,106 @@ private void AddAllAdminFacets()
+                _state = StateDestroyInProgress;
+            }
+
+            // Cancel operations that are waiting and using the communicator's cancellation token
+            _cancellationTokenSource.Cancel();
+
+            // Shutdown and destroy all the incoming and outgoing Ice connections and wait for the connections to be
+            // finished.
+            _ = ShutdownAsync();
+
+            _outgoingConnectionFactory?.DestroyAsync();
+
+            await WaitForShutdownAsync().ConfigureAwait(false);
+            _shutdownTaskSemaphore?.Dispose();
+
+            // At this point adapters are immutable
+            foreach (ObjectAdapter adapter in _adapters)
+            {
+                await adapter.DisposeAsync().ConfigureAwait(false);
+            }
+
+            // _adminAdapter is disposed above when iterating over all adapters, we call DisposeAsync here to avoid the
+            // compiler warning about disposable field not being dispose.
+            if (_adminAdapter != null)
+            {
+                await _adminAdapter.DisposeAsync().ConfigureAwait(false);
+            }
+
+            if (_outgoingConnectionFactory != null)
+            {
+                await _outgoingConnectionFactory.DestroyAsync().ConfigureAwait(false);
+            }
+
+            Observer?.SetObserverUpdater(null);
+
+            if (Logger is LoggerAdminLogger adminLogger)
+            {
+                await adminLogger.DisposeAsync().ConfigureAwait(false);
+            }
+
+            if (GetPropertyAsBool("Ice.Warn.UnusedProperties") ?? false)
+            {
+                List<string> unusedProperties = GetUnusedProperties();
+                if (unusedProperties.Count != 0)
+                {
+                    var message = new StringBuilder("The following properties were set but never read:");
+                    foreach (string s in unusedProperties)
+                    {
+                        message.Append("\n    ");
+                        message.Append(s);
+                    }
+                    Logger.Warning(message.ToString());
+                }
+            }
+
+            // Destroy last so that a Logger plugin can receive all log/traces before its destruction.
+            List<(string Name, IPlugin Plugin)> plugins;
+            lock (_mutex)
+            {
+                plugins = new List<(string Name, IPlugin Plugin)>(_plugins);
+            }
+            plugins.Reverse();
+            foreach ((string name, IPlugin plugin) in plugins)
+            {
+                try
+                {
+                    await plugin.DisposeAsync().ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    Runtime.Logger.Warning($"unexpected exception raised by plug-in `{name}' destruction:\n{ex}");
+                }
+            }
+
+            lock (_mutex)
+            {
+                _state = StateDestroyed;
+            }
```

Comment:
We should consider getting rid of `_state`. Shouldn't the methods that check for _state == Destroyed instead check for `_disposeTask != null`?

Reply:
I get rid of Communicator _state, but still keep _isShutdown.